src/main/scala/com/fulcrumgenomics/alignment/Aligner.scala

```diff
@@ -49,61 +48,65 @@ object Mode extends FgBioEnum[Mode] {
+  private val LeftAndDiagonal: Seq[Direction] = Seq(Diagonal, Left)
+  private val UpAndDiagonal: Seq[Direction]   = Seq(Diagonal, Up)
 
   /** The minimum score allowed to start an alignment.  This prevents underflow. */
   val MinStartScore: Int = Int.MinValue / 2
-}
-
 
-object AlignmentMatrix {
-  def apply(direction: Direction, queryLength: Int, targetLength: Int): AlignmentMatrix = {
-    AlignmentMatrix(direction=direction, scoring=Matrix[Int](queryLength+1, targetLength+1), trace=Matrix[Direction](queryLength+1, targetLength+1))
+  object AlignmentMatrix {
+    def apply(direction: Direction, queryLength: Int, targetLength: Int): AlignmentMatrix = {
+      AlignmentMatrix(direction=direction, scoring=Matrix[Int](queryLength+1, targetLength+1), trace=Matrix[Direction](queryLength+1, targetLength+1))
+    }
   }
-}
 
-/** A single alignment matrix for a given [[Direction]] storing both the scoring and traceback matrices produce by the aligner. */
-case class AlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction])
+  /** A single alignment matrix for a given [[Direction]] storing both the scoring and traceback matrices produce by the aligner. */
+  case class AlignmentMatrix(direction: Direction, scoring: Matrix[Int], trace: Matrix[Direction])
+}
 
 
 /**
-  * Implementation of the Needleman-Wunsch algorithm for global alignment of two sequences with support for an
-  * affine gap penalty.
+  * Implementation of an aligner with generic scoring function and affine gap penalty support.
+  * Supports multiple alignment [[Mode]]s for global, semi-global and local alignment.
+  *
+  * A scoring function (`scoringFunction`) is taken to score pair-wise aligned bases. A default
+  * implementation is supplied via the companion [[Aligner]] object which uses a fixed match
+  * score and mismatch penalty.
   *
-  * @param scoringFunction a function to score the alignment of a pair of bases
-  * @param gapOpen the gap opening penalty
-  * @param gapExtend the gap extension penalty
+  * @param scoringFunction a function to score the alignment of a pair of bases. Parameters
+  *                        are the query and target bases as bytes, in that order.
+  * @param gapOpen the gap opening penalty, should generally be negative or zero
+  * @param gapExtend the gap extension penalty, should generally be negative or zero
   * @param useEqualsAndX if true use the = and X cigar operators for matches and mismatches,
   *                      else use the M operator for both.
   * @param mode alignment mode to use when generating alignments
   */
-class NeedlemanWunschAligner(val scoringFunction: (Byte,Byte) => Int,
-                             val gapOpen: Int,
-                             val gapExtend: Int,
-                             useEqualsAndX: Boolean = true,
-                             val mode: Mode = Global) {
+class Aligner(val scoringFunction: (Byte,Byte) => Int,
+              val gapOpen: Int,
```

Comment:
Question for you @nh13.  Do you think there's ever a valid case for providing `gapOpen` or `gapExtend` > 0?  I frequently get confused as some aligners take the penalties as positive numbers and subtract them, and some (as we do) take negative numbers and add them.  I'd like to add a pair of:

```scala
require(gapOpen <= 0, "...")
require(gapExtend <= 0, "...")
```

but wanted to see if you thought that was a bad idea for any reason?